2020-06-16: Add post security-audits (#320)
oktadeveloper/okta-blog: _source/_posts/2020-06-16-security-audits.adoc

In the Java world, I'm a big fan of https://spotbugs.github.io/[SpotBugs] and https://find-sec-bugs.github.io/[Find Security Bugs]. No tool is going to find all bugs, but these tools have found more issues than I would have expected and more importantly, have taught me a great deal about various types of bugs and how to avoid them.

In-depth code analysis is great, but when it comes from outside your team, they are generally slow and too costly (time and money) to do against every commit. This often leaves teams to do them "as needed." The catch 22 is—if you are not a security expert—how do you know when they are needed? One option is to schedule them periodically—quarterly, for example—or some other frequency that matches the pace of your development.

== Outside Audit

For the first release of https://github.com/paseto-toolkit/jpaseto[JPaseto] we hired https://paragonie.com/[Paragon Initiative Enterprises] to do a security audit. They had domain expertise, as they helped author the https://paseto.io/rfc/[PASETO RFC]. The turn around was fast, and they came back with a few issues:

. Constant Time Comparison
+

> This was the most critical of the issues they found. The library was making a comparison of a hashed value using Java's `Arrays.equals()` method which fails fast, because it's optimized for speed. The code should have been using a time constant comparison such as `MessageDigest.isEqual()` to prevent timing attacks.

Reviewer: I just learned about the + line in AsciiDoc a few weeks ago! It's pretty slick and can be used to add code blocks into lists too.

Author: Nice! I didn't know that!

. Variable Names
+
Typically, you don't think of variable names when talking about security issues, but an incorrectly named variable can lead to maintenance issues or improper usage in the future. In this particular case, there was a variable named `rsaSignature` that but was an EdDSA signature.

> The fix is obvious: rename the variable to `eddsaSignature`.

Reviewer: Obvious? Why not `edDsaSignature`? :trollface:

Author: good point!
do you think "obvious" is too strong? Maybe something like:
```suggestion
The fix is simple: rename the variable to `edDsaSignature `.
```

Reviewer: Yeah, I think it's a bit strong.

> Overall my experience with Paragon Initiative was great. They even found a typo in the project's README.md. A doc issue doesn't sound like much, but it shows their attention to detail while going through a codebase.

Reviewer: ```suggestion
Overall my experience with Paragon Initiative was great. They even found a typo in the project's `README.md`. A doc issue doesn't sound like much, but it shows their attention to detail while going through a codebase.
```

== Vulnerabilities Happen

As software developers, we know no code is bug-free, but for some reason, we think things will be different for security issues. This makes no sense because we know security is hard, and we know bugs are inevitable. Even after running automated tools on your codebase and getting a professional audit, there is still a likelihood of finding other issues. Of course, that is what happened.

I'll use a simplified example to point out the problem, take a look at the following code, and see if you can spot it;

[source,java]
----
public interface BasicMath {
    int add(int a, int b);
    int subtract(int b, int a);
    int multiply(int a, int b);
    int divide(int a, int b);
}
----

> In Java, the variable order maters. It's important to make sure all your methods use the same order. In the above example, the `subtract` methods arguments are `b, a` all of the other methods are ordered `a, b`. In isolation, this doesn' really matter, and the implementation might look like this:

Reviewer: ```suggestion
In Java, the variable order matters. It's important to make sure all your methods use the same order. In the above example, the `subtract` methods arguments are `b, a` all of the other methods are ordered `a, b`. In isolation, this doesn't really matter, and the implementation might look like this:
```

[source,java]
----
public class ExampleBasicMath {
public int add(int a, int b) {
return a + b;

> }

Reviewer: This code block looks like it needs to be formatted.
